### ISSUE TITLE
Crash with `std::type_info`/`typeid` (`CXXTypeidExpr`)

### ISSUE THREAD

**Issue by jspam** [mgehre/clang#35]
When compiling the following source code with `-Wlifetime`, Clang crashes on my machine:

```cpp
#include <typeinfo>

class A
{
};

void Foo(const std::type_info&) { };

void Bar() {
    Foo(typeid(A));
}
```

Version information:
clang `53964d1643e9e1250e0cf72ceca9c731a4e7a902`
llvm `d8e175bca5919de13fe003cab6b10f23af3d055d`
g++ `g++-8 (Debian 8.1.0-1) 8.1.0` (for system includes)

clang and llvm were built in RelWithDebInfo mode.

Console output/stack trace:
```
CXXTypeidExpr 0x55555b15b250 'const class std::type_info' lvalue
Missing entry in RefersTo
UNREACHABLE executed at /home/user/xdata/hdd/llvm/tools/clang/lib/Analysis/LifetimePsetBuilder.cpp:1125!
Stack dump:
0.	Program arguments: /mnt/xdata/hdd/llvm-build-500/bin/clang-7.0 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all -disable-free -main-file-name LifetimeTest.cpp -mrelocation-model static -mthread-model posix -mdisable-fp-elim -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -dwarf-column-info -debugger-tuning=gdb -resource-dir /mnt/xdata/hdd/llvm-build-500/lib/clang/7.0.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/x86_64-linux-gnu/c++/8 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/x86_64-linux-gnu/c++/8 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/backward -internal-isystem /usr/local/include -internal-isystem /mnt/xdata/hdd/llvm-build-500/lib/clang/7.0.0/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -Wlifetime -Wlifetime-debug -std=c++17 -fdeprecated-macro -fdebug-compilation-dir /home/user -ferror-limit 19 -fmessage-length 211 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -fcolor-diagnostics -o /tmp/LifetimeTest-edb15e.o -x c++ LifetimeTest.cpp 
1.	<eof> parser at end of file
2.	LifetimeTest.cpp:9:12: parsing function body 'Bar'
#0 0x0000555556c52f5a llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/user/xdata/hdd/llvm/lib/Support/Unix/Signals.inc:494:0
#1 0x0000555556c513d4 llvm::sys::RunSignalHandlers() /home/user/xdata/hdd/llvm/lib/Support/Signals.cpp:67:0
#2 0x0000555556c51502 SignalHandler(int) /home/user/xdata/hdd/llvm/lib/Support/Unix/Signals.inc:353:0
#3 0x00007ffff7bcd0c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x110c0)
#4 0x00007ffff6762fff gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x32fff)
#5 0x00007ffff676442a abort (/lib/x86_64-linux-gnu/libc.so.6+0x3442a)
#6 0x0000555556c04eda (/mnt/xdata/hdd/llvm-build-500/bin/clang-7.0+0x16b0eda)
#7 0x000055555867a530 clang::lifetime::PSetsBuilder::VisitBlock(clang::CFGBlock const&, llvm::Optional<std::map<clang::lifetime::Variable, clang::lifetime::PSet, std::less<clang::lifetime::Variable>, std::allocator<std::pair<clang::lifetime::Variable const, clang::lifetime::PSet> > > >&) /home/user/xdata/hdd/llvm/tools/clang/lib/Analysis/LifetimePsetBuilder.cpp:1128:0
#8 0x000055555867a6cd clang::lifetime::VisitBlock(std::map<clang::lifetime::Variable, clang::lifetime::PSet, std::less<clang::lifetime::Variable>, std::allocator<std::pair<clang::lifetime::Variable const, clang::lifetime::PSet> > >&, llvm::Optional<std::map<clang::lifetime::Variable, clang::lifetime::PSet, std::less<clang::lifetime::Variable>, std::allocator<std::pair<clang::lifetime::Variable const, clang::lifetime::PSet> > > >&, clang::lifetime::PSet const&, std::map<clang::Expr const*, clang::lifetime::PSet, std::less<clang::Expr const*>, std::allocator<std::pair<clang::Expr const* const, clang::lifetime::PSet> > >&, std::map<clang::Expr const*, clang::lifetime::PSet, std::less<clang::Expr const*>, std::allocator<std::pair<clang::Expr const* const, clang::lifetime::PSet> > >&, clang::CFGBlock const&, clang::lifetime::LifetimeReporterBase&, clang::ASTContext&, llvm::function_ref<bool (clang::QualType, clang::QualType)>) /home/user/xdata/hdd/llvm/tools/clang/lib/Analysis/LifetimePsetBuilder.cpp:1186:0
#9 0x0000555558667a39 clang::lifetime::LifetimeContext::TraverseBlocks() /home/user/xdata/hdd/llvm/tools/clang/lib/Analysis/Lifetime.cpp:198:0
#10 0x00005555586687d9 _ZNSt8_Rb_treeIPKN5clang4ExprESt4pairIKS3_NS0_8lifetime4PSetEESt10_Select1stIS8_ESt4lessIS3_ESaIS8_EED4Ev /usr/include/c++/7/bits/stl_tree.h:949:0
#11 0x00005555586687d9 _ZNSt3mapIPKN5clang4ExprENS0_8lifetime4PSetESt4lessIS3_ESaISt4pairIKS3_S5_EEED4Ev /usr/include/c++/7/bits/stl_map.h:294:0
#12 0x00005555586687d9 _ZN5clang8lifetime15LifetimeContextD4Ev /home/user/xdata/hdd/llvm/tools/clang/lib/Analysis/Lifetime.cpp:28:0
#13 0x00005555586687d9 clang::lifetime::runAnalysis(clang::FunctionDecl const*, clang::ASTContext&, clang::lifetime::LifetimeReporterBase&, llvm::function_ref<bool (clang::QualType, clang::QualType)>, llvm::function_ref<clang::FunctionDecl* (clang::CXXRecordDecl const*, clang::OverloadedOperatorKind)>, llvm::function_ref<clang::FunctionDecl* (clang::CXXRecordDecl const*, llvm::StringRef)>, llvm::function_ref<void (clang::ClassTemplateSpecializationDecl*)>) /home/user/xdata/hdd/llvm/tools/clang/lib/Analysis/Lifetime.cpp:233:0
#14 0x000055555861952c clang::sema::AnalysisBasedWarnings::IssueWarnings(clang::sema::AnalysisBasedWarnings::Policy, clang::sema::FunctionScopeInfo*, clang::Decl const*, clang::BlockExpr const*) /home/user/xdata/hdd/llvm/tools/clang/lib/Sema/AnalysisBasedWarnings.cpp:2399:0
#15 0x00005555580bf837 clang::Sema::PopFunctionScopeInfo(clang::sema::AnalysisBasedWarnings::Policy const*, clang::Decl const*, clang::BlockExpr const*) /home/user/xdata/hdd/llvm/tools/clang/lib/Sema/Sema.cpp:1411:0
#16 0x00005555581b9474 clang::Sema::ActOnFinishFunctionBody(clang::Decl*, clang::Stmt*, bool) /home/user/xdata/hdd/llvm/tools/clang/lib/Sema/SemaDecl.cpp:12924:0
#17 0x0000555557f1c919 clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) /home/user/xdata/hdd/llvm/tools/clang/lib/Parse/ParseStmt.cpp:1969:0
#18 0x0000555557e91d5c clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) /home/user/xdata/hdd/llvm/tools/clang/lib/Parse/Parser.cpp:1243:0
#19 0x0000555557eb6107 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::SourceLocation*, clang::Parser::ForRangeInit*) /home/user/xdata/hdd/llvm/tools/clang/lib/Parse/ParseDecl.cpp:1953:0
#20 0x0000555557e8ee18 clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) /home/user/xdata/hdd/llvm/tools/clang/lib/Parse/Parser.cpp:1013:0
#21 0x0000555557e8f3fa _ZN5clang6Parser21ObjCDeclContextSwitchD4Ev /home/user/xdata/hdd/llvm/tools/clang/include/clang/Parse/Parser.h:892:0
#22 0x0000555557e8f3fa clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) /home/user/xdata/hdd/llvm/tools/clang/lib/Parse/Parser.cpp:1026:0
#23 0x0000555557e9307f clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) /home/user/xdata/hdd/llvm/tools/clang/lib/Parse/Parser.cpp:853:0
#24 0x0000555557e941ae clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) /home/user/xdata/hdd/llvm/tools/clang/lib/Parse/Parser.cpp:610:0
#25 0x0000555557e8af67 clang::ParseAST(clang::Sema&, bool, bool) /home/user/xdata/hdd/llvm/tools/clang/lib/Parse/ParseAST.cpp:157:0
#26 0x00005555576011d8 clang::CodeGenAction::ExecuteAction() /home/user/xdata/hdd/llvm/tools/clang/lib/CodeGen/CodeGenAction.cpp:1044:0
#27 0x000055555728266e clang::FrontendAction::Execute() /home/user/xdata/hdd/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:910:0
#28 0x0000555557246cae clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/user/xdata/hdd/llvm/tools/clang/lib/Frontend/CompilerInstance.cpp:962:0
#29 0x0000555557320af1 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/user/xdata/hdd/llvm/tools/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:257:0
#30 0x0000555555f32580 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/user/xdata/hdd/llvm/tools/clang/tools/driver/cc1_main.cpp:216:0
#31 0x0000555555edc9f6 ExecuteCC1Tool /home/user/xdata/hdd/llvm/tools/clang/tools/driver/driver.cpp:310:0
#32 0x0000555555edc9f6 main /home/user/xdata/hdd/llvm/tools/clang/tools/driver/driver.cpp:382:0
#33 0x00007ffff67502e1 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202e1)
#34 0x0000555555f2eaaa _start (/mnt/xdata/hdd/llvm-build-500/bin/clang-7.0+0x9daaaa)
clang-7.0: error: unable to execute command: Aborted
clang-7.0: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 7.0.0 (http://llvm.org/git/clang 53964d1643e9e1250e0cf72ceca9c731a4e7a902) (http://llvm.org/git/llvm d8e175bca5919de13fe003cab6b10f23af3d055d)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /mnt/xdata/hdd/llvm-build-500/bin
```

I can provide the preprocessed source code if needed.

**Comment by mgehre** [comment]
Fixed in 99451dc6bfb8fcf48a371fb390f9d9f7e19be779

**Comment by mgehre** [comment]
If you have any other feedback or experience, please let us know!
